quarkus/jib: Native build fails if docker isn't installed

### Describe the bug

1) Download https://github.com/operator-framework/operator-sdk
2) mkdir test && cd test
3) operator-sdk init --plugins quarkus --domain halkyon.io --project-name expose
4) Change pom.xml
```
    <quarkus-sdk.version>3.0.4</quarkus-sdk.version>
    <quarkus.version>2.7.3.Final</quarkus.version>
```
5) `mvn clean package -Pnative`

Fails with

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.7.3.Final:build (default) on project expose: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.container.image.jib.deployment.JibProcessor#buildFromNative threw an exception: java.lang.RuntimeException: Unable to create container image
[ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.containerize(JibProcessor.java:236)
[ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.buildFromNative(JibProcessor.java:203)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:882)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:832)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.util.concurrent.ExecutionException: java.io.IOException: Cannot run program "docker": error=2, No such file or directory
[ERROR] 	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:566)
[ERROR] 	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:547)
[ERROR] 	at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:88)
[ERROR] 	at com.google.cloud.tools.jib.builder.steps.StepsRunner.run(StepsRunner.java:219)
[ERROR] 	at com.google.cloud.tools.jib.api.Containerizer.run(Containerizer.java:390)
[ERROR] 	at com.google.cloud.tools.jib.api.JibContainerBuilder.containerize(JibContainerBuilder.java:598)
[ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.containerize(JibProcessor.java:229)
[ERROR] 	... 12 more
[ERROR] Caused by: java.io.IOException: Cannot run program "docker": error=2, No such file or directory
[ERROR] 	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128)
[ERROR] 	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071)
[ERROR] 	at com.google.cloud.tools.jib.docker.DockerClient.docker(DockerClient.java:277)
[ERROR] 	at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:193)
[ERROR] 	at com.google.cloud.tools.jib.builder.steps.LoadDockerStep.call(LoadDockerStep.java:74)
[ERROR] 	at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$loadDocker$18(StepsRunner.java:618)
[ERROR] 	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
[ERROR] 	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
[ERROR] 	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
[ERROR] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
[ERROR] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:832)
[ERROR] Caused by: java.io.IOException: error=2, No such file or directory
[ERROR] 	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
[ERROR] 	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:319)
[ERROR] 	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:250)
[ERROR] 	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107)
[ERROR] 	... 11 more
```

since `docker` isn't installed. The build up to this point have used `podman` successfully.

### Expected behavior

To be able to use `podman` for the entire build

### Actual behavior

`docker` is required - e.g. `podman` isn't detected

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

JDK14

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.7.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.4

### Additional information

Workaround: Using

```
mvn clean package -Dquarkus.docker.executable-name=podman -Pnative
```

and the build is successful.

Code: extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java
@@ -2,10 +2,10 @@ package io.quarkus.deployment.pkg;
 
 import java.io.File;
 import java.util.List;
-import java.util.Locale;
 import java.util.Optional;
 import java.util.OptionalInt;
 
+import io.quarkus.deployment.util.ContainerRuntimeUtil;
 import io.quarkus.runtime.annotations.ConfigGroup;
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;
@@ -229,7 +229,7 @@ public class NativeConfig {
      * a container build is always done.
      */
     @ConfigItem
-    public Optional<ContainerRuntime> containerRuntime;
+    public Optional<ContainerRuntimeUtil.ContainerRuntime> containerRuntime;
 
     /**
      * Options to pass to the container runtime
@@ -440,18 +440,6 @@ public class NativeConfig {
          */
         @ConfigItem
         public Optional<List<String>> additionalArgs;
-    }
-
-    /**
-     * Supported Container runtimes
-     */
-    public static enum ContainerRuntime {
-        DOCKER,
-        PODMAN;
-
-        public String getExecutableName() {
-            return this.name().toLowerCase();
-        }
     }
 
     /**

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildContainerRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildContainerRunner.java
@@ -1,7 +1,6 @@
 package io.quarkus.deployment.pkg.steps;
 
 import java.io.IOException;
-import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -15,7 +14,7 @@ import org.apache.commons.lang3.RandomStringUtils;
 import org.jboss.logging.Logger;
 
 import io.quarkus.deployment.pkg.NativeConfig;
-import io.quarkus.deployment.util.FileUtil;
+import io.quarkus.deployment.util.ContainerRuntimeUtil;
 import io.quarkus.deployment.util.ProcessUtil;
 
 public abstract class NativeImageBuildContainerRunner extends NativeImageBuildRunner {
@@ -23,14 +22,14 @@ public abstract class NativeImageBuildContainerRunner extends NativeImageBuildRu
     private static final Logger log = Logger.getLogger(NativeImageBuildContainerRunner.class);
 
     final NativeConfig nativeConfig;
-    protected final NativeConfig.ContainerRuntime containerRuntime;
+    protected final ContainerRuntimeUtil.ContainerRuntime containerRuntime;
     String[] baseContainerRuntimeArgs;
     protected final String outputPath;
     private final String containerName;
 
     public NativeImageBuildContainerRunner(NativeConfig nativeConfig, Path outputDir) {
         this.nativeConfig = nativeConfig;
-        containerRuntime = nativeConfig.containerRuntime.orElseGet(NativeImageBuildContainerRunner::detectContainerRuntime);
+        containerRuntime = nativeConfig.containerRuntime.orElseGet(ContainerRuntimeUtil::detectContainerRuntime);
         log.infof("Using %s to run the native image builder", containerRuntime.getExecutableName());
 
         this.baseContainerRuntimeArgs = new String[] { "--env", "LANG=C", "--rm" };
@@ -41,8 +40,8 @@ public abstract class NativeImageBuildContainerRunner extends NativeImageBuildRu
 
     @Override
     public void setup(boolean processInheritIODisabled) {
-        if (containerRuntime == NativeConfig.ContainerRuntime.DOCKER
-                || containerRuntime == NativeConfig.ContainerRuntime.PODMAN) {
+        if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.DOCKER
+                || containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
             // we pull the docker image in order to give users an indication of which step the process is at
             // it's not strictly necessary we do this, however if we don't the subsequent version command
             // will appear to block and no output will be shown
@@ -123,53 +122,6 @@ public abstract class NativeImageBuildContainerRunner extends NativeImageBuildRu
                 .of(Stream.of(containerRuntime.getExecutableName()), Stream.of(dockerCmd), Stream.of(baseContainerRuntimeArgs),
                         containerRuntimeArgs.stream(), Stream.of(nativeConfig.getEffectiveBuilderImage()), command.stream())
                 .flatMap(Function.identity()).toArray(String[]::new);
-    }
-
-    /**
-     * @return {@link NativeConfig.ContainerRuntime#DOCKER} if it's available, or {@link NativeConfig.ContainerRuntime#PODMAN}
-     *         if the podman
-     *         executable exists in the environment or if the docker executable is an alias to podman
-     * @throws IllegalStateException if no container runtime was found to build the image
-     */
-    public static NativeConfig.ContainerRuntime detectContainerRuntime() {
-        // Docker version 19.03.14, build 5eb3275d40
-        String dockerVersionOutput = getVersionOutputFor(NativeConfig.ContainerRuntime.DOCKER);
-        boolean dockerAvailable = dockerVersionOutput.contains("Docker version");
-        // Check if Podman is installed
-        // podman version 2.1.1
-        String podmanVersionOutput = getVersionOutputFor(NativeConfig.ContainerRuntime.PODMAN);
-        boolean podmanAvailable = podmanVersionOutput.startsWith("podman version");
-        if (dockerAvailable) {
-            // Check if "docker" is an alias to "podman"
-            if (dockerVersionOutput.equals(podmanVersionOutput)) {
-                return NativeConfig.ContainerRuntime.PODMAN;
-            }
-            return NativeConfig.ContainerRuntime.DOCKER;
-        } else if (podmanAvailable) {
-            return NativeConfig.ContainerRuntime.PODMAN;
-        } else {
-            throw new IllegalStateException("No container runtime was found to run the native image builder. "
-                    + "Make sure you have Docker or Podman installed in your environment.");
-        }
-    }
-
-    private static String getVersionOutputFor(NativeConfig.ContainerRuntime containerRuntime) {
-        Process versionProcess = null;
-        try {
-            ProcessBuilder pb = new ProcessBuilder(containerRuntime.getExecutableName(), "--version")
-                    .redirectErrorStream(true);
-            versionProcess = pb.start();
-            versionProcess.waitFor();
-            return new String(FileUtil.readFileContents(versionProcess.getInputStream()), StandardCharsets.UTF_8);
-        } catch (IOException | InterruptedException e) {
-            // If an exception is thrown in the process, just return an empty String
-            log.debugf(e, "Failure to read version output from %s", containerRuntime.getExecutableName());
-            return "";
-        } finally {
-            if (versionProcess != null) {
-                versionProcess.destroy();
-            }
-        }
     }
 
 }

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
@@ -11,6 +11,7 @@ import java.util.List;
 import org.apache.commons.lang3.SystemUtils;
 
 import io.quarkus.deployment.pkg.NativeConfig;
+import io.quarkus.deployment.util.ContainerRuntimeUtil;
 import io.quarkus.deployment.util.FileUtil;
 
 public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContainerRunner {
@@ -23,7 +24,7 @@ public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContai
             String gid = getLinuxID("-gr");
             if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
                 Collections.addAll(containerRuntimeArgs, "--user", uid + ":" + gid);
-                if (containerRuntime == NativeConfig.ContainerRuntime.PODMAN) {
+                if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
                     // Needed to avoid AccessDeniedExceptions
                     containerRuntimeArgs.add("--userns=keep-id");
                 }

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/UpxCompressionBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/UpxCompressionBuildStep.java
@@ -20,6 +20,7 @@ import io.quarkus.deployment.pkg.NativeConfig;
 import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
 import io.quarkus.deployment.pkg.builditem.NativeImageBuildItem;
 import io.quarkus.deployment.pkg.builditem.UpxCompressedBuildItem;
+import io.quarkus.deployment.util.ContainerRuntimeUtil;
 import io.quarkus.deployment.util.FileUtil;
 import io.quarkus.deployment.util.ProcessUtil;
 
@@ -99,8 +100,8 @@ public class UpxCompressionBuildStep {
         List<String> extraArgs = nativeConfig.compression.additionalArgs.orElse(Collections.emptyList());
 
         List<String> commandLine = new ArrayList<>();
-        NativeConfig.ContainerRuntime containerRuntime = nativeConfig.containerRuntime
-                .orElseGet(NativeImageBuildContainerRunner::detectContainerRuntime);
+        ContainerRuntimeUtil.ContainerRuntime containerRuntime = nativeConfig.containerRuntime
+                .orElseGet(ContainerRuntimeUtil::detectContainerRuntime);
         commandLine.add(containerRuntime.getExecutableName());
 
         commandLine.add("run");
@@ -121,7 +122,7 @@ public class UpxCompressionBuildStep {
             String gid = getLinuxID("-gr");
             if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
                 Collections.addAll(commandLine, "--user", uid + ":" + gid);
-                if (containerRuntime == NativeConfig.ContainerRuntime.PODMAN) {
+                if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
                     // Needed to avoid AccessDeniedExceptions
                     commandLine.add("--userns=keep-id");
                 }

--- a/core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java
@@ -1,0 +1,73 @@
+package io.quarkus.deployment.util;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+import org.jboss.logging.Logger;
+
+public final class ContainerRuntimeUtil {
+
+    private static final Logger log = Logger.getLogger(ContainerRuntimeUtil.class);
+
+    private ContainerRuntimeUtil() {
+    }
+
+    /**
+     * @return {@link ContainerRuntime#DOCKER} if it's available, or {@link ContainerRuntime#PODMAN}
+     *         if the podman
+     *         executable exists in the environment or if the docker executable is an alias to podman
+     * @throws IllegalStateException if no container runtime was found to build the image
+     */
+    public static ContainerRuntime detectContainerRuntime() {
+        // Docker version 19.03.14, build 5eb3275d40
+        String dockerVersionOutput = getVersionOutputFor(ContainerRuntime.DOCKER);
+        boolean dockerAvailable = dockerVersionOutput.contains("Docker version");
+        // Check if Podman is installed
+        // podman version 2.1.1
+        String podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
+        boolean podmanAvailable = podmanVersionOutput.startsWith("podman version");
+        if (dockerAvailable) {
+            // Check if "docker" is an alias to "podman"
+            if (dockerVersionOutput.equals(podmanVersionOutput)) {
+                return ContainerRuntime.PODMAN;
+            }
+            return ContainerRuntime.DOCKER;
+        } else if (podmanAvailable) {
+            return ContainerRuntime.PODMAN;
+        } else {
+            throw new IllegalStateException("No container runtime was found to. "
+                    + "Make sure you have Docker or Podman installed in your environment.");
+        }
+    }
+
+    private static String getVersionOutputFor(ContainerRuntime containerRuntime) {
+        Process versionProcess = null;
+        try {
+            ProcessBuilder pb = new ProcessBuilder(containerRuntime.getExecutableName(), "--version")
+                    .redirectErrorStream(true);
+            versionProcess = pb.start();
+            versionProcess.waitFor();
+            return new String(FileUtil.readFileContents(versionProcess.getInputStream()), StandardCharsets.UTF_8);
+        } catch (IOException | InterruptedException e) {
+            // If an exception is thrown in the process, just return an empty String
+            log.debugf(e, "Failure to read version output from %s", containerRuntime.getExecutableName());
+            return "";
+        } finally {
+            if (versionProcess != null) {
+                versionProcess.destroy();
+            }
+        }
+    }
+
+    /**
+     * Supported Container runtimes
+     */
+    public enum ContainerRuntime {
+        DOCKER,
+        PODMAN;
+
+        public String getExecutableName() {
+            return this.name().toLowerCase();
+        }
+    }
+}

--- a/extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java
+++ b/extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java
@@ -69,6 +69,7 @@ import io.quarkus.deployment.pkg.builditem.OutputTargetBuildItem;
 import io.quarkus.deployment.pkg.builditem.UpxCompressedBuildItem;
 import io.quarkus.deployment.pkg.steps.JarResultBuildStep;
 import io.quarkus.deployment.pkg.steps.NativeBuild;
+import io.quarkus.deployment.util.ContainerRuntimeUtil;
 import io.quarkus.fs.util.ZipUtils;
 import io.quarkus.maven.dependency.ResolvedDependency;
 
@@ -266,6 +267,11 @@ public class JibProcessor {
                 dockerDaemonImage.setDockerExecutable(Paths.get(jibConfigExecutableName.get()));
             } else if (dockerConfigExecutableName.isPresent()) {
                 dockerDaemonImage.setDockerExecutable(Paths.get(dockerConfigExecutableName.get()));
+            } else {
+                // detect the container runtime instead of falling back to 'docker' as the default
+                ContainerRuntimeUtil.ContainerRuntime detectedContainerRuntime = ContainerRuntimeUtil.detectContainerRuntime();
+                log.infof("Using %s to run the native image builder", detectedContainerRuntime.getExecutableName());
+                dockerDaemonImage.setDockerExecutable(Paths.get(detectedContainerRuntime.getExecutableName()));
             }
             containerizer = Containerizer.to(dockerDaemonImage);
         }

